### PR TITLE
Added additional distance metrics for k-NN: Jira-1059

### PR DESCRIPTION
**Pull request by hpandeycodeit**

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/285/

File: src/ports/postgres/modules/knn/knn.py_in [hunk 1]
```diff
@@ -88,12 +88,28 @@ def knn_validate_src(schema_madlib, point_source, point_column_name, point_id,
         plpy.error("kNN Error: Data type '{0}' is not a valid type for"
                    " column '{1}' in table '{2}'.".
                    format(col_type_test, test_id, test_source))
+
+    fn_dist = fn_dist.lower().strip()
```
Comment (iyerr3) [inline src/ports/postgres/modules/knn/knn.py_in]: What's the expected behavior if `fn_dist = None` (i.e. user inputs `NULL` for the `fn_dist`)?

File: src/ports/postgres/modules/knn/test/knn.sql_in [hunk 2]
```diff
@@ -73,23 +73,23 @@ copy knn_test_data (id, data) from stdin delimiter '|';
...
 select assert(array_agg(prediction order by id)='{1,1,0.5,1,0.25,0.25}', 'Wrong output in regression') from madlib_knn_result_regression;
 
 drop table if exists madlib_knn_result_regression;
-select knn('knn_train_data_reg','data','id','label','knn_test_data','data','id','madlib_knn_result_regression',3,True);
+select knn('knn_train_data_reg','data','id','label','knn_test_data','data','id','madlib_knn_result_regression',3,True,'madlib.squared_dist_norm2');
 select assert(array_agg(x)= '{1,2,3}' , 'Wrong output in regression with k=3') from (select unnest(k_nearest_neighbours) as x from madlib_knn_result_regression where id = 1 order by x asc) y;
 
 drop table if exists madlib_knn_result_classification;
-select knn('knn_train_data','data','id','label','knn_test_data','data','id','madlib_knn_result_classification',False);
+select knn('knn_train_data','data','id','label','knn_test_data','data','id','madlib_knn_result_classification',3,False,'madlib.squared_dist_norm2');
```
Comment (iyerr3) [inline src/ports/postgres/modules/knn/test/knn.sql_in]: How about we add other dist functions (including `NULL` input) in above functions?

File: src/ports/postgres/modules/knn/knn.py_in (hunk 1 above)
Comment (hpandeycodeit) [reply src/ports/postgres/modules/knn/knn.py_in]: I have added a check here. If fn_dist is None, set it to default 'squared_dist_norm2' function.

File: src/ports/postgres/modules/knn/knn.py_in [hunk 3]
```diff
@@ -88,12 +88,28 @@ def knn_validate_src(schema_madlib, point_source, point_column_name, point_id,
...
+    dist_functions = {
```
[the remark this answers is not in the document]
Comment (hpandeycodeit) [reply src/ports/postgres/modules/knn/knn.py_in]: This is changed as suggested.

File: src/ports/postgres/modules/knn/knn.py_in [hunk 4]
```diff
@@ -88,12 +88,28 @@ def knn_validate_src(schema_madlib, point_source, point_column_name, point_id,
...
+        schema_madlib + '.dist_norm1',
+        schema_madlib + '.dist_norm2',
+        schema_madlib + '.squared_dist_norm2',
+        schema_madlib + '.dist_angle',
+        schema_madlib + '.dist_tanimoto'}
+
+    if plpy.execute("""select prorettype != 'DOUBLE PRECISION'::regtype
```
[the remark this answers is not in the document]
Comment (hpandeycodeit) [reply src/ports/postgres/modules/knn/knn.py_in]: This is changed as suggested.

File: src/ports/postgres/modules/knn/test/knn.sql_in (hunk 2 above)
Comment (hpandeycodeit) [reply src/ports/postgres/modules/knn/test/knn.sql_in]: Added few more test functions where fn_dist is passed as NULL. All the above changes are committed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/286/

**Comment by fmcquillan99** [comment]
@hpandeycodeit we would like to get this PR in the upcoming 1.13 release.  Are you planning to do additional work as per the comments above?

File: src/ports/postgres/modules/knn/knn.py_in [hunk 5]
```diff
@@ -89,20 +89,20 @@ def knn_validate_src(schema_madlib, point_source, point_column_name, point_id,
...
+        dist_functions = set([schema_madlib + dist for dist in
+                              ('.dist_norm1', '.dist_norm2', '.squared_dist_norm2', '.dist_angle', '.dist_tanimoto')])
+
+        is_invalid_func = plpy.execute(
+            """select prorettype != 'DOUBLE PRECISION'::regtype
+                OR proisagg = TRUE AS OUTPUT from pg_proc where
+                oid='{fn_dist}(DOUBLE PRECISION[], DOUBLE PRECISION[])'::regprocedure;
+                """.format(**locals()))[0]['output']
+
+        if is_invalid_func or fn_dist not in dist_functions:
```
[the remark this answers is not in the document]
Comment (hpandeycodeit) [reply src/ports/postgres/modules/knn/knn.py_in]: @orhankislal @njayaram2 @kaknikhil @fmcquillan99 Does the user input has to be in this form Schema_name.Fn_dist ? I referred to the k-Means where  Schema_name.Fn_dist is passed as the user input.

File: src/ports/postgres/modules/knn/knn.py_in [hunk 6]
```diff
@@ -89,20 +89,20 @@ def knn_validate_src(schema_madlib, point_source, point_column_name, point_id,
...
-        schema_madlib + '.dist_angle',
-        schema_madlib + '.dist_tanimoto'}
-
-    if plpy.execute("""select prorettype != 'DOUBLE PRECISION'::regtype
-            OR proisagg = TRUE AS OUTPUT from pg_proc where
-            oid='{fn_dist}(DOUBLE PRECISION[], DOUBLE PRECISION[])'::regprocedure;"""
-                    .format(**locals()))[0]['output'] or fn_dist not in dist_functions:
-        plpy.error(
-            "KNN error: Distance function has wrong signature or is not a simple function.")
+    if fn_dist and fn_dist is not None:
```
[the remark this answers is not in the document]
Comment (hpandeycodeit) [reply src/ports/postgres/modules/knn/knn.py_in]: Changed this as well.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/289/

**Comment by iyerr3** [comment]
Changes look good. I'll merge this.
